debuginfod/src/resolver.rs

```diff
@@ -0,0 +1,159 @@
+use dirs::cache_dir;
+use std::fs;
+use std::fs::File;
+use std::path::PathBuf;
+
+use crate::client::DebugInfodClient;
+
+use blazesym::symbolize::Elf;
+use blazesym::symbolize::Source;
+
+use anyhow::anyhow;
+use anyhow::Result;
+
+#[derive(Debug)]
+pub struct DebugInfodResolver {
+    ignore_cache: bool,
+    base_path: PathBuf,
+    clients: Vec<DebugInfodClient>,
+}
+
+impl DebugInfodResolver {
+    /// Returns a DebugInfodResolver resolver.
+    pub fn new(
+        clients: Vec<DebugInfodClient>,
+        base_path: PathBuf,
+        ignore_cache: bool,
+    ) -> Result<DebugInfodResolver> {
+        Ok(DebugInfodResolver {
+            ignore_cache: ignore_cache,
+            base_path: base_path,
+            clients: clients,
+        })
+    }
+
+    pub fn cache_dir() -> Result<PathBuf> {
+        let mut dir = cache_dir()
+            .or(Some(std::env::temp_dir()))
+            .expect("Could not create temp dir");
+        dir.push("blazesym");
+        dir.push("debuginfod");
+        Ok(dir)
+    }
+
+    /// Returns a default DebugInfodResolver resolver suitable for use. Uses DEBUGINFOD_URLS for
+    /// configuring debuginfod clients. Caches resolved objects by default.
+    pub fn default_resolver() -> Result<DebugInfodResolver> {
+        let clients = DebugInfodClient::get_default_clients()?;
+        let cache_dir = DebugInfodResolver::cache_dir()?;
+        std::fs::create_dir_all(cache_dir.clone())?;
```

Comment:
Untested, but I don't think we have to clone.
```suggestion
        std::fs::create_dir_all(&cache_dir)?;
````